airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftCopier.java

```diff
@@ -0,0 +1,213 @@
+      throws IOException {
+    this(RedshiftCopyDestination.DEFAULT_AIRBYTE_STAGING_S3_BUCKET, runFolder, syncMode, schema, streamName, client, database, s3KeyId, s3key,
+        s3Region);
+  }
+
+  @VisibleForTesting
+  public RedshiftCopier(
+                        String s3BucketName,
+                        String runFolder,
+                        SyncMode syncMode,
+                        String schema,
+                        String streamName,
+                        AmazonS3 client,
+                        JdbcDatabase database,
+                        String s3KeyId,
+                        String s3key,
+                        String s3Region)
+      throws IOException {
+    this.s3BucketName = s3BucketName;
+    this.runFolder = runFolder;
+    this.syncMode = syncMode;
+    this.schemaName = schema;
+    this.streamName = streamName;
+    this.s3Client = client;
+    this.database = database;
+    this.s3KeyId = s3KeyId;
+    this.s3Key = s3key;
+    this.s3Region = s3Region;
+
+    // The stream transfer manager lets us greedily stream into S3. The native AWS SDK does not
+    // have support for streaming multipart uploads;
+    // The alternative is first writing the entire output to disk before loading into S3. This is not
+    // feasible with large tables.
+    // Data is chunked into parts. A part is sent off to a queue to be uploaded once it has reached it's
+    // configured part size.
+    // Memory consumption is queue capacity * part size = 10 * 10 = 100 MB at current configurations.
+    this.multipartUploadManager =
+        new StreamTransferManager(s3BucketName, getPath(runFolder, streamName), client)
+            .numUploadThreads(DEFAULT_UPLOAD_THREADS)
+            .queueCapacity(DEFAULT_QUEUE_CAPACITY)
+            .partSize(PART_SIZE_MB);
+    // We only need one output stream as we only have one input stream. This is reasonably performant.
+    // See the above comment.
+    this.outputStream = multipartUploadManager.getMultiPartOutputStreams().get(0);
+
+    var writer = new PrintWriter(outputStream, true, StandardCharsets.UTF_8);
+    this.csvPrinter = new CSVPrinter(writer, CSVFormat.DEFAULT);
+  }
+
+  public void copy(AirbyteRecordMessage message) throws IOException {
+    var id = UUID.randomUUID();
+    var data = Jsons.serialize(message.getData());
+    var emittedAt = Timestamp.from(Instant.ofEpochMilli(message.getEmittedAt()));
+    csvPrinter.printRecord(id, data, emittedAt);
+  }
+
+  public void close(boolean hasFailed) throws IOException {
+    // TODO (dchia): revisit error handling here
+    if (hasFailed) {
+      multipartUploadManager.abort();
+      return;
+    }
+    closeS3WriteStreamAndUpload();
+
+    var tmpTableName = NAMING_RESOLVER.getTmpTableName(streamName);
+    var sqlOp = new RedshiftSqlOperations();
+    try {
+      createTmpTableAndCopyS3FileInto(tmpTableName, sqlOp);
+      mergeIntoDestTableIncrementalOrFullRefresh(tmpTableName, sqlOp);
```

Comment:
is there a way to do this in a single transaction for all tables? if not then we need to document expectations in the error cases. currently for other destinations, we are able to do this copy atomically.

Reply:
great point. nothing jumps out at me for now; will play around and change this.